File: thesync-backend/model/schedule.py
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import PositiveInt, field_validator, model_validator

from model.base import DomainModel, NonEmptyText, normalize_optional_text


class Schedule(DomainModel):
    """Domain and API representation of a schedule request."""

    id: UUID
    student_id: UUID
    adviser_id: UUID
    type_id: PositiveInt
    status_id: PositiveInt
    topic: NonEmptyText
    requested_at: datetime
    scheduled_at: datetime | None = None
    google_calendar_event_id: str | None = None
    meet_link: str | None = None
    created_at: datetime

    @field_validator("google_calendar_event_id", "meet_link", mode="before")
    @classmethod
    def normalize_optional_strings(cls, value: str | None) -> str | None:
        return normalize_optional_text(value)

    @model_validator(mode="after")
    def validate_users(self) -> Schedule:
        if self.student_id == self.adviser_id:
            raise ValueError("student_id and adviser_id must reference different users.")

        return self


class ScheduleListFilters(DomainModel):
    """Filters and pagination for schedule list queries."""

    status_id: PositiveInt | None = None
    type_id: PositiveInt | None = None
    from_date: date | datetime | None = None
    to_date: date | datetime | None = None
    page: PositiveInt = 1
    page_size: PositiveInt = 20


class ScheduleCreateRequest(DomainModel):
    """Payload for creating a schedule request."""

    adviser_id: UUID
    type_id: PositiveInt
    topic: NonEmptyText
    scheduled_at: datetime | None = None


class ScheduleApproveRequest(DomainModel):
    """Payload for approving a schedule."""

    scheduled_at: datetime | None = None


class ScheduleRejectRequest(DomainModel):
    """Payload for rejecting a schedule."""

    remarks: str | None = None

    @field_validator("remarks", mode="before")
    @classmethod
    def normalize_rejection_remarks(cls, value: str | None) -> str | None:
        return normalize_optional_text(value)


class ScheduleRescheduleRequest(DomainModel):
    """Payload for rescheduling a schedule."""

    scheduled_at: datetime
    remarks: str | None = None

    @field_validator("remarks", mode="before")
    @classmethod
    def normalize_reschedule_remarks(cls, value: str | None) -> str | None:
        return normalize_optional_text(value)

File: thesync-frontend/types/schedule.ts
/** Minimal user fields embedded inside a ScheduleDetail response */
export interface UserSummary {
  id: string;
  full_name: string;
}

/** Panelist assignment entry with resolved names (for ScheduleDetail) */
export interface PanelistDetail {
  id: string;
  panelist_id: string;
  full_name: string;
  invite_status: string;
}

/**
 * Full schedule detail response for GET /schedules/{id}
 *
 * Differences from Schedule:
 * - student/adviser are embedded UserSummary objects instead of raw UUIDs
 * - type and status are resolved name strings
 * - panelist_assignments contains resolved names
 * - meet_link is only populated when status == approved
 * - google_calendar_event_id is only populated for admin/adviser callers
 */
export interface ScheduleDetail {
  id: string;
  topic: string;
  student: UserSummary;
  adviser: UserSummary;
  type: string;
  status: string;
  requested_at: string;
  scheduled_at: string | null;
  meet_link: string | null;
  google_calendar_event_id: string | null;
  panelist_assignments: PanelistDetail[];
  created_at: string;
}

/** Domain and API representation of a schedule request */
export interface Schedule {
  id: string;
  student_id: string;
  adviser_id: string;
  type_id: number;
  status_id: number;
  topic: string;
  requested_at: string;
  scheduled_at: string | null;
  google_calendar_event_id: string | null;
  meet_link: string | null;
  created_at: string;
}

/** Expanded schedule row returned by list endpoints */
export interface ScheduleListItem extends Schedule {
  student_full_name: string;
  adviser_full_name: string;
  type_name: string;
  status_name: string;
}

/** Paginated schedule list response */
export interface ScheduleListResponse {
  items: ScheduleListItem[];
  total: number;
  page: number;
  limit: number;
}

/** Payload for creating a schedule request */
export interface ScheduleCreateRequest {
  adviser_id: string;
  type_id: number;
  topic: string;
  scheduled_at: string;
}

/** Payload for approving a schedule */
export interface ScheduleApproveRequest {
  scheduled_at?: string | null;
}

/** Payload for rejecting a schedule */
export interface ScheduleRejectRequest {
  remarks?: string | null;
}

/** Payload for rescheduling a schedule */
export interface ScheduleRescheduleRequest {
  scheduled_at: string;
  remarks?: string | null;
}
